oauth2-code-flow-sample/server.js

```diff
@@ -0,0 +1,115 @@
+import path from "path";
+import dotenv from "dotenv";
+import express from "express";
+import { Issuer } from "openid-client";
+import cookieParser from "cookie-parser";
+
+dotenv.config();
+
+const clientId = process.env.CLIENT_ID;
+const clientSecret = process.env.CLIENT_SECRET;
+const redirectUri = process.env.REDIRECT_URI;
+const scopes = process.env.SCOPES;
+const port = process.env.PORT || 3000;
+
+const redirectPath = new URL(redirectUri).pathname;
+
+/**
+ * Creates and returns a new OpenID Connect client with the provided client credentials.
+ */
+const createClient = async () => {
+  const issuer = await Issuer.discover("https://volvoid.eu.qa.volvocars.com");
+
+  const client = new issuer.Client({
+    client_id: clientId,
+    client_secret: clientSecret,
+    redirect_uris: [redirectUri],
+    response_types: ["code"],
+  });
+
+  return client;
+};
+
+const main = async () => {
+  const app = express();
+  const __dirname = path.resolve();
+
+  app.use(express.json());
+  app.use(express.static(__dirname + "/public"));
+  app.use(cookieParser());
+
+  const client = await createClient();
+
+  // Renders the main view
+  app.get("/", (req, res) => {
+    const { refresh_token, access_token } = req.cookies;
+
+    if (access_token && refresh_token) {
+      res.sendFile(path.join(__dirname, "main.html"));
+    } else {
+      res.redirect("/login");
+    }
+  });
+
+  // Renders the login view
+  app.get("/login", (req, res) => {
+    res.sendFile(path.join(__dirname, "login.html"));
+  });
+
+  /**
+   * This endpoint is the destination for Volvo ID after a successful login, also known as your redirect URI.
+   * It will provide the login code as a query parameter and exchange it for access and refresh tokens.
+   **/
+  app.get(redirectPath, async (req, res) => {
+    const code = req.query.code;
+
+    if (code) {
+      const tokenSet = await client.callback(redirectUri, {
+        code,
+        grant_type: "authorization_code",
+      });
+
+      res.cookie("refresh_token", tokenSet.refresh_token);
+      res.cookie("access_token", tokenSet.access_token);
+
+      res.redirect("/");
+    } else {
+      res.redirect("/login");
```

Comment:
Maybe log the error here? If something was wrong with the login an error message will be added as a query parameter.

Know we said, "no error handing", but in this case, I think logging the error will help users see what is wrong in their redirects.

Reply:
Good point! I'll add